File: leetcode/dp/322. Coin Change/py/main.py
# Time O(S * N)
# Space O(N)
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        inf = float('inf')
        dp = [inf] * (amount + 1)

        dp[0] = 0

        for target in range(1, amount + 1):
          for i in range(len(coins)):
            if target - coins[i] >= 0:
                dp[target] = min(dp[target], dp[target - coins[i]] + 1)
                
        return -1 if dp[amount] == inf else dp[amount] 

File: leetcode/dp/322. Coin Change/js/main.js
// Time O(S * N). На каждом шаге алгоритм находит следующий F (i) за N итераций, где 1 <= i <= S.
// Следовательно, всего итераций S * N итераций
// Space O(N)
const coinChange = (coins, amount) => {
  let dp = Array(amount + 1).fill(amount + 1);
  dp[0] = 0;

  for (let i = 1; i <= amount; i++) {
    for (let j = 0; j < coins.length; j++) {
      if (i < coins[j]) continue;

      dp[i] = Math.min(dp[i - coins[j]] + 1, dp[i]);
    }
  }

  return dp[amount] !== amount + 1 ? dp[amount] : -1;
};
